### ISSUE TITLE
Use `SymInteger` instead of `SymNat` to represent regions for `LLVMPointer`

### ISSUE THREAD

**Issue by danmatichuk** [GaloisInc/crucible#1027]
Now that `SymNat` is a wrapped `SymInteger` with a positivity check, it is not clear that there is any benefit to using them to represent regions for pointers in the LLVM memory model. `SymNat` is awkward to work with, as it requires a special case in nearly all code that is otherwise generic over What4 base types.

Switching `LLVMPointer` to using `SymInteger` would likely just be an exercise in deleting all of the special cases that were introduced when `SymNat` was added as a separate type.

**Comment by RyanGlScott** [comment]
I'm not quite sure I understand what the benefit of switching to `SymInteger` is. If anything, it seems like it would incur the downside of no longer being able to assure that the region number is non-negative. Perhaps I haven't encountered the awkwardness of using `SymNat` that you're alluding to—can you provide an example?

**Comment by travitch** [comment]
Nothing relies on the non-negativity. We only ever test block IDs for equality (and usually against 0)

**Comment by RyanGlScott** [comment]
Fair enough. I'm still not clear on what the awkwardness of using `SymNat` is, but at the same time, I don't have a compelling argument for its continued use in `LLVMPointer`.

**Comment by travitch** [comment]
I think @danmatichuk's main complaint is that each time you project from a `SymNat` to a base type (`SymInteger`), it wraps the term in a `max 0 x` (where `x`) is the value you actually care about. These terms stack up over time if you start doing things with them and make things unreadable (and may also cause problems if you lose some context that makes the term easily reducible).

**Comment by RyanGlScott** [comment]
Ah, that is a very good point. If we're not making use of the ordering of region numbers, then I agree that switching to `SymInteger` would be an improvement in that regard.

**Comment by travitch** [comment]
I believe this is the function that causes particular pain: https://github.com/GaloisInc/what4/blob/7b9f4563de467c9fdfe1f26f4f0fbd1f2c03d829/what4/src/What4/Interface.hs#L534

**Comment by danmatichuk** [comment]
Indeed one issue is that mapping to and from `SymNat`s causes a buildup of redundant muxes, specifically when the abstract domain of a term is weakened at some point and the positivity check can't be simplified away.

An alternative/additional solution would be to relax `What4`s abstract domain interface a bit and allow users to manually set the domain for any expression. This would allow the `SymNat` interface to be a bit more robust, since it could retain the positivity of the unwrapped integer in its abstract domain.

**Comment by travitch** [comment]
You can already (unsafely, obviously) set the abstract domain for a value: https://github.com/GaloisInc/what4/blob/7b9f4563de467c9fdfe1f26f4f0fbd1f2c03d829/what4/src/What4/Interface.hs#L382

**Comment by danmatichuk** [comment]
Aha, somehow I missed this. I think if SymNat set the abstract domain when unwrapping integers it might actually mitigate a lot of these issues. I'll have to do some testing to see how much it helps the shape of the terms.

I still think moving to SymInteger is a good idea, simply due to the reduction in code complexity, but perhaps less pressing.

**Comment by travitch** [comment]
I like the idea of using this (relatively new) primitive to improve the behavior of our existing nat<->int primitives is a great intermediate step. Definitely interested to hear if that is effective.

**Comment by danmatichuk** [comment]
Here is my proposed fix, which should ensure that the use of `intMax` in `SymNat` uses and propagates as much abstract domain information as possible.

https://github.com/GaloisInc/what4/pull/216

With this change, we can more easily control when the inner integer of a `SymNat` gets wrapped in a positivity check by setting the abstract domain, and integers that come from unwrapping a `SymNat` should be statically known to be necessarily positive and therefore not wrapped in any additional checks.

**Comment by langston-barrett** [comment]
See also https://github.com/GaloisInc/crucible/issues/404
